woocommerce/storefront
# Basic e2e test setup using wp-env, jest and puppeteer [in progress]

Related #1330 - this is the first step, a few follow up tasks mentioned here.

The goal here is to get a basic e2e setup so that we can incrementally add tests for front-end & admin dashboard flows. This will provide an early warning when PRs impact existing flows (catch regressions) and another automated way to test releases. Also as we build more tests this will provide a good reference for what flows are supported in Storefront.

This PR adds basic config for the following:

- A [wp-env](https://developer.wordpress.org/block-editor/packages/packages-env/) test container (server) environment, for running tests against. Could be used as a standard/default dev setup for the repo too potentially.
- Jest & Puppeteer and related dependencies. This is the combination used by other projects, notably WooCommerce core. Jest is the test runner and Puppeteer allows tests to target UI flows in a browser - aka end-to-end (e2e) tests.
- A very basic test to confirm that the `Built with Storefront` footer text is present 😁

#### What needs doing?
- ~~The store url is hard-coded - need to configure this. @senadir can you advise how to do this? I see in WooCommerce Blocks there's a `default.json` config file with env vars, but I don't know how to tell Jest/puppeteer about such a file.~~ Update: now using [jest `globals`](https://jestjs.io/docs/en/configuration#globals-object), this works well. However I'm still interested to hear other / better approaches.
- Docs for how to run tests and contribute new tests, and how it all works.
- Explore and clarify our needs re WooCommerce and WordPress test utils. As we start implementing tests for real flows we will want to leverage a lib for navigating admin dashboard and other common flows not implemented by us. None of this is in place yet - TBD if this is a blocker for this PR.
- Longer term, I'd love to be able to run core WooCommerce and WordPress flows as part of our test suite. Examples: commenting on a post, or add-to-cart and complete purchase. I suspect that we'll be able to write these tests by using utils libs, am curious if we can share more of these flows tests. 


### How to test the changes in this Pull Request:
1. Get storefront built: clone storefront repo, `npm install`, `composer install`, `npm run build`.
2. Start wp-env environment: `wp-env start` from repo folder. 
4. When it's complete you should be able to see woo+storefront site at http://localhost:8888 - note that this is using the default ports. If you are running another env you'll need to take it down (`wp-env stop`). 
3. Run the tests: `npm run e2e` from repo folder.

Should output happy test result:

<img width="560" alt="Screen Shot 2020-06-18 at 4 56 13 PM" src="https://user-images.githubusercontent.com/4167300/84979882-a9282000-b184-11ea-90a4-19e6e702d56b.png">

Note about ports - I'm keen to set up Storefront on custom ports so it doesn't clash with other projects using the default, but I hit an issue with that - https://github.com/WordPress/gutenberg/issues/23259

### Changelog

> Dev - Add support for end-to-end tests using Jest & Puppeteer. #1394
